File: src/main/java/dev/obscuria/obscure_tooltips/client/tooltip/filter/EnchantmentFilter.java
package dev.obscuria.obscure_tooltips.client.tooltip.filter;

import com.github.bsideup.jabel.Desugar;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import net.minecraft.enchantment.Enchantment;
import net.minecraft.enchantment.EnchantmentHelper;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.nbt.NBTTagList;
import net.minecraft.util.JsonUtils;
import net.minecraft.util.ResourceLocation;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

@Desugar
public record EnchantmentFilter(Boolean anyEnchantment, Boolean anyCurse, List<ResourceLocation> enchantments)
        implements ItemFilter {

    public static EnchantmentFilter fromJson(JsonObject json) {
        final Boolean anyEnchantment = JsonUtils.hasField(json, "any_enchantment")
                ? JsonUtils.getBoolean(json, "any_enchantment") : null;
        final Boolean anyCurse = JsonUtils.hasField(json, "any_curse")
                ? JsonUtils.getBoolean(json, "any_curse") : null;
        List<ResourceLocation> enchantments = null;
        if (JsonUtils.hasField(json, "enchantments")) {
            enchantments = new ArrayList<>();
            for (JsonElement element : JsonUtils.getJsonArray(json, "enchantments")) {
                enchantments.add(new ResourceLocation(element.getAsString()));
            }
        }
        return new EnchantmentFilter(anyEnchantment, anyCurse, enchantments);
    }

    @Override
    public boolean test(ItemStack stack) {
        if (anyEnchantment != null && anyEnchantment != stack.isItemEnchanted()) {
            return false;
        }
        if (anyCurse != null && anyCurse != isCursed(stack)) {
            return false;
        }
        return enchantments == null || containsAll(stack, enchantments);
    }

    private static boolean isCursed(ItemStack stack) {
        return readEnchantments(stack, "ench").stream().anyMatch(EnchantmentFilter::isCurse)
                || readEnchantments(stack, "StoredEnchantments").stream().anyMatch(EnchantmentFilter::isCurse);
    }

    private static boolean containsAll(ItemStack stack, List<ResourceLocation> enchantmentIds) {
        final List<Enchantment> present = new ArrayList<>(EnchantmentHelper.getEnchantments(stack).keySet());
        return enchantmentIds.stream()
                .map(Enchantment.REGISTRY::getObject)
                .allMatch(present::contains);
    }

    private static boolean isCurse(Enchantment enchantment) {
        final ResourceLocation id = Enchantment.REGISTRY.getNameForObject(enchantment);
        return id != null && id.getPath().toLowerCase(Locale.ROOT).contains("curse");
    }

    private static List<Enchantment> readEnchantments(ItemStack stack, String tagName) {
        final List<Enchantment> result = new ArrayList<>();
        final NBTTagCompound tag = stack.getTagCompound();
        if (tag == null || !tag.hasKey(tagName, 9)) { // 9 = TAG_LIST
            return result;
        }
        final NBTTagList list = tag.getTagList(tagName, 10); // 10 = TAG_COMPOUND
        for (int i = 0; i < list.tagCount(); i++) {
            final Enchantment enchantment = Enchantment.getEnchantmentByID(list.getCompoundTagAt(i).getShort("id"));
            if (enchantment != null) {
                result.add(enchantment);
            }
        }
        return result;
    }
}

File: common/src/main/kotlin/dev/obscuria/tooltips/client/filter/EnchantmentFilter.kt
package dev.obscuria.tooltips.client.filter

import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.enchantment.Enchantment
import net.minecraft.world.item.enchantment.EnchantmentHelper
import java.util.Optional
import kotlin.jvm.optionals.getOrNull

class EnchantmentFilter(
    val anyEnchantment: Optional<Boolean>,
    val anyCurse: Optional<Boolean>,
    val enchantments: Optional<List<Enchantment>>
) : IItemFilter {

    override fun codec(): Codec<EnchantmentFilter> = CODEC

    override fun test(stack: ItemStack): Boolean {
        if (anyEnchantment.getOrNull()?.equals(stack.isEnchanted) == false) return false
        if (anyCurse.getOrNull()?.equals(stack.isCursed()) == false) return false
        if (enchantments.getOrNull()?.containsIn(stack) == false) return false
        return true
    }

    companion object {

        val CODEC: Codec<EnchantmentFilter> = RecordCodecBuilder.create { codec ->
            codec.group(
                Codec.BOOL
                    .optionalFieldOf("any_enchantment")
                    .forGetter(EnchantmentFilter::anyEnchantment),
                Codec.BOOL
                    .optionalFieldOf("any_curse")
                    .forGetter(EnchantmentFilter::anyCurse),
                BuiltInRegistries.ENCHANTMENT.byNameCodec().listOf()
                    .optionalFieldOf("enchantments")
                    .forGetter(EnchantmentFilter::enchantments)
            ).apply(codec, ::EnchantmentFilter)
        }

        fun ItemStack.isCursed(): Boolean {
            return EnchantmentHelper.getEnchantments(this).any { it.key.isCurse }
        }

        fun List<Enchantment>.containsIn(stack: ItemStack): Boolean {
            return this.all { EnchantmentHelper.getItemEnchantmentLevel(it, stack) > 0 }
        }
    }
}